docs/rotation.md

```diff
@@ -0,0 +1,175 @@
+
+```bash
+kubectl logs -l component=azure-kms-provider -n kube-system
+
+I0219 17:35:33.608840       1 main.go:60] "Starting KeyManagementServiceServer service" version="v0.0.11" buildDate="2021-02-19-17:33"
+I0219 17:35:33.609090       1 azure_config.go:27] populating AzureConfig from /etc/kubernetes/azure.json
+I0219 17:35:33.609420       1 auth.go:66] "azure: using client_id+client_secret to retrieve access token" clientID="9a7a##### REDACTED #####bb26" clientSecret="23T.##### REDACTED #####vw-r"
+I0219 17:35:33.609568       1 keyvault.go:66] "using kms key for encrypt/decrypt" vaultName="k8skmskv" keyName="key1" keyVersion="5cdf48ea6bb9456ebf637e1130b7751a"
+I0219 17:35:33.609897       1 main.go:86] Listening for connections on address: /opt/azurekms2.socket
+...
+```
+
+### 3. Add the new provider to encryption configuration in `/etc/kubernetes/manifests/encryptionconfig.yaml`
+
+```yaml
+kind: EncryptionConfiguration
+apiVersion: apiserver.config.k8s.io/v1
+resources:
+    - resources:                                            # List of kubernetes resources that will be encrypted in etcd using the KMS plugin
+        - secrets
+    providers:
+        - kms:
+            name: azurekmsprovider
+            endpoint: unix:///opt/azurekms.socket           # This endpoint must match the value defined in --listen-addr for the KMS plugin using old key
+            cachesize: 1000
+        - kms:
+            name: azurekmsprovider2
+            endpoint: unix:///opt/azurekms2.socket          # This endpoint must match the value defined in --listen-addr for the KMS plugin using new key
+            cachesize: 1000
+```
+
+### 4. Restart all `kube-apiserver`
+
+* Proceed to the next step if using a single `kube-apiserver`
+* If using multi-master, restart the `kube-apiserver` to ensure each server can still decrypt using the new key in the encryption config.
+  * To validate the decryption still works, run `kubectl get secret <secret name> -o yaml` with one of the existing secrets to confirm the data is returned and is valid.
+
+### 5. Switch the order of provider in the encryption config
+
+```yaml
+kind: EncryptionConfiguration
+apiVersion: apiserver.config.k8s.io/v1
+resources:
+    - resources:                                            # List of kubernetes resources that will be encrypted in etcd using the KMS plugin
+        - secrets
+    providers:
+        # kms provider with new key
+        - kms:
+            name: azurekmsprovider2
+            endpoint: unix:///opt/azurekms2.socket          # This endpoint must match the value defined in --listen-addr for the KMS plugin using new key
+            cachesize: 1000
+        # kms provider with old key
+        - kms:
+            name: azurekmsprovider
+            endpoint: unix:///opt/azurekms.socket           # This endpoint must match the value defined in --listen-addr for the KMS plugin using old key
+            cachesize: 1000
+```
+
+### 6. Restart all `kube-apiserver`
```

Comment:
same as number 4?

Reply:
do they need to refer to same instructions in number 4?

Reply:
Yeah, added a reference to step 4.